### PR TITLE
[nrf noup] BLE Extended Advertisement option (#31532)

### PR DESCRIPTION
**Pull request by LipinskiPNordicSemi**
This commit implements solution for new config CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING which allow device to be visibly longer for the commissioning.

Cherry-picked PR content:

* additions to support changes as discussed in connectedhomeip-spec #7825

* add Silabs platform example implementation

* add extension for ExtendedAnnouncementFlag + Silabs platform usage

* Restyled by clang-format

* adjust based on review comments - 1

* Restyled by clang-format

* add missing include guard for default case

* Revert "Restyled by clang-format"

This reverts commit 6d3f9e107de9a8421524437f44c2113730e52e54.

* fix extra bracket causing a bug + restyle changes

* Restyled by clang-format

* adjust based on review comments - 2

* Restyled by clang-format

---------

### CODE REVIEW

File: src/include/platform/CHIPDeviceConfig.h [hunk 1]
```diff
@@ -693,7 +693,11 @@ static_assert(CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING_INTERVAL_MIN < CHIP_DEVICE_
  * Time in seconds that a factory new device will advertise commissionable node discovery.
  */
 #ifndef CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_SECS
+#if CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING == 0
```
Comment (LuDuda) [inline src/include/platform/CHIPDeviceConfig.h:696]: Can we please introduce the Zephyr's KConfig for this feature?

File: src/platform/Zephyr/BLEManagerImpl.cpp [hunk 2]
```diff
@@ -363,10 +396,17 @@ CHIP_ERROR BLEManagerImpl::StartAdvertising()
 
         if (mFlags.Has(Flags::kFastAdvertisingEnabled))
         {
-            // Start timer to change advertising interval.
+            // Start timer to change advertising fast to slow interval.
             DeviceLayer::SystemLayer().StartTimer(
                 System::Clock::Milliseconds32(CHIP_DEVICE_CONFIG_BLE_ADVERTISING_INTERVAL_CHANGE_TIME),
-                HandleBLEAdvertisementIntervalChange, this);
+                HandleSlowBLEAdvertisementInterval, this);
+#if CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
```
Comment (LuDuda) [inline src/platform/Zephyr/BLEManagerImpl.cpp:403]: Move one line below?

File: src/platform/Zephyr/BLEManagerImpl.cpp [hunk 3]
```diff
@@ -60,6 +60,8 @@ namespace {
 constexpr uint32_t kAdvertisingOptions = BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_ONE_TIME;
 constexpr uint8_t kAdvertisingFlags    = BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR;
 
+constexpr uint8_t kLowestRequestPriority = 7;
```
Comment (Damian-Nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:63]: Unused?

File: src/platform/Zephyr/BLEManagerImpl.cpp [hunk 4]
```diff
@@ -293,18 +299,45 @@ inline CHIP_ERROR BLEManagerImpl::PrepareAdvertisingRequest()
...
+        else
+        {
+            mAdvertisingRequest.minInterval = CHIP_DEVICE_CONFIG_BLE_SLOW_ADVERTISING_INTERVAL_MIN;
+            mAdvertisingRequest.maxInterval = CHIP_DEVICE_CONFIG_BLE_SLOW_ADVERTISING_INTERVAL_MAX;
+        }
+#else
+        mAdvertisingRequest.minInterval = CHIP_DEVICE_CONFIG_BLE_SLOW_ADVERTISING_INTERVAL_MIN;
+        mAdvertisingRequest.maxInterval = CHIP_DEVICE_CONFIG_BLE_SLOW_ADVERTISING_INTERVAL_MAX;
+#endif
+    }
```
Comment (Damian-Nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:340]:
```suggestion
#if CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
      else if (mFlags.Has(Flags::kExtendedAdvertisingEnabled))
      {
          mAdvertisingRequest.minInterval = CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING_INTERVAL_MIN;
          mAdvertisingRequest.maxInterval = CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING_INTERVAL_MAX;
      }
#endif
      else
      {
          mAdvertisingRequest.minInterval = CHIP_DEVICE_CONFIG_BLE_SLOW_ADVERTISING_INTERVAL_MIN;
          mAdvertisingRequest.maxInterval = CHIP_DEVICE_CONFIG_BLE_SLOW_ADVERTISING_INTERVAL_MAX;
      }
```

File: src/include/platform/CHIPDeviceConfig.h (hunk 1 above)
Comment (Damian-Nordic) [reply src/include/platform/CHIPDeviceConfig.h:696]: option: we could probably have a Kconfig for advertising duration and turn on extended advertising if the duration > 15 minutes.

File: src/platform/Zephyr/BLEManagerImpl.cpp [hunk 5]
```diff
@@ -245,7 +247,11 @@ void BLEManagerImpl::DriveBLEState()
     }
     else
     {
-        if (mFlags.Has(Flags::kAdvertising))
+        if (mFlags.Has(Flags::kAdvertising)
+#if CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
+            && mFlags.Has(Flags::kExtendedAdvertisingEnabled)
```
Comment (Damian-Nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:252]: Can you explain this change? Won't this cause that if the commissioning window is somehow closed earlier by a user application, the advertising won't be stopped?

File: src/platform/Zephyr/BLEManagerImpl.cpp [hunk 6]
```diff
@@ -423,6 +464,9 @@ CHIP_ERROR BLEManagerImpl::_SetAdvertisingMode(BLEAdvertisingMode mode)
     case BLEAdvertisingMode::kSlowAdvertising:
         mFlags.Set(Flags::kFastAdvertisingEnabled, false);
         break;
+    case BLEAdvertisingMode::kExtendedAdvertising:
+        mFlags.Set(Flags::kExtendedAdvertisingEnabled, true);
```
Comment (Damian-Nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:468]: I think you should clear this flag somewhere so that slow advertising is not broken when the commissioning window is started again.

File: src/platform/Zephyr/BLEManagerImpl.cpp (hunk 5 above)
Comment (LipinskiPNordicSemi) [reply src/platform/Zephyr/BLEManagerImpl.cpp:252]: My bad, I misunderstood the part of this code

File: src/include/platform/CHIPDeviceConfig.h (hunk 1 above)
Comment (LuDuda) [reply src/include/platform/CHIPDeviceConfig.h:696]: I would also go with single config  (ext adv) or alternatively two configs (with timeout). The numbers 15 minutes and 48h are defined in the spec, so while configurability always is desired.. here it may result e.g. in failing the certification test case.

File: src/app/server/CommissioningWindowManager.h [hunk 7]
```diff
@@ -59,8 +59,13 @@ class CommissioningWindowManager : public Messaging::UnsolicitedMessageHandler,
 
     static constexpr System::Clock::Seconds16 MaxCommissioningTimeout()
     {
+#if CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
+        // Specification section 2.3.1 - Extended Announcement Duration up to 48h
+        return System::Clock::Seconds16(60 * 60 * 48);
```
Comment (ArekBalysNordic) [inline src/app/server/CommissioningWindowManager.h:64]: I know that it comes from upstream, but are you sure that the resulting number doesn't overflow `Seconds16`? What is the actual impact on our code?

File: src/include/platform/CHIPDeviceConfig.h (hunk 1 above)
Comment (Damian-Nordic) [reply src/include/platform/CHIPDeviceConfig.h:696]: The spec only mentions maximum duration of the announcement and that after 15 minutes it should switch to the extended one. Thus, a manufacturer can decide to advertise e.g. 1 hour if I read the spec correctly.

File: src/app/server/CommissioningWindowManager.h (hunk 7 above)
Comment (LipinskiPNordicSemi) [reply src/app/server/CommissioningWindowManager.h:64]: Yeah, just realized it after @kkasperczyk-no previous comment.

File: src/include/platform/CHIPDeviceConfig.h (hunk 1 above)
Comment (Damian-Nordic) [reply src/include/platform/CHIPDeviceConfig.h:696]: Though, of course, we can go with a single boolean Kconfig if you think that will be easier. A customer can use chip_project_config.h to modify the duration, anyway.

File: src/platform/Zephyr/BLEManagerImpl.cpp (hunk 6 above)
Comment (markaj-nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:468]: Apart from clearing `kExtendedAdvertisingEnabled` somewhere, should we clear other flags in this `case`?

File: src/platform/Zephyr/BLEManagerImpl.cpp [hunk 8]
```diff
@@ -363,10 +396,17 @@ CHIP_ERROR BLEManagerImpl::StartAdvertising()
...
         {
-            // Start timer to change advertising interval.
+            // Start timer to change advertising fast to slow interval.
             DeviceLayer::SystemLayer().StartTimer(
                 System::Clock::Milliseconds32(CHIP_DEVICE_CONFIG_BLE_ADVERTISING_INTERVAL_CHANGE_TIME),
-                HandleBLEAdvertisementIntervalChange, this);
+                HandleSlowBLEAdvertisementInterval, this);
+#if CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
+
+            // Start timer to change advertising slow to extended interval.
```
Comment (markaj-nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:405]: ...to change advertising interval from slow to extended

File: src/platform/Zephyr/BLEManagerImpl.cpp [hunk 9]
```diff
@@ -363,10 +396,17 @@ CHIP_ERROR BLEManagerImpl::StartAdvertising()
 
         if (mFlags.Has(Flags::kFastAdvertisingEnabled))
         {
-            // Start timer to change advertising interval.
+            // Start timer to change advertising fast to slow interval.
```
Comment (markaj-nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:399]: ...to change advertising interval from fast to slow

File: src/platform/Zephyr/BLEManagerImpl.cpp [hunk 10]
```diff
@@ -293,18 +299,45 @@ inline CHIP_ERROR BLEManagerImpl::PrepareAdvertisingRequest()
     Encoding::LittleEndian::Put16(serviceData.uuid, UUID16_CHIPoBLEService.val);
     ReturnErrorOnFailure(ConfigurationMgr().GetBLEDeviceIdentificationInfo(serviceData.deviceIdInfo));
 
+#if CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
+    if (mFlags.Has(Flags::kExtendedAdvertisingEnabled))
+    {
+        serviceData.deviceIdInfo.SetVendorId(DEVICE_HANDLE_NULL);
```
Comment (markaj-nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:305]: Why don't we set the proper device info?

File: src/platform/Zephyr/BLEManagerImpl.cpp [hunk 11]
```diff
@@ -293,18 +299,45 @@ inline CHIP_ERROR BLEManagerImpl::PrepareAdvertisingRequest()
...
+    {
+#if CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
+        if (mFlags.Has(Flags::kExtendedAdvertisingEnabled))
+        {
+            mAdvertisingRequest.minInterval = CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING_INTERVAL_MIN;
+            mAdvertisingRequest.maxInterval = CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING_INTERVAL_MAX;
+        }
+        else
+        {
+            mAdvertisingRequest.minInterval = CHIP_DEVICE_CONFIG_BLE_SLOW_ADVERTISING_INTERVAL_MIN;
```
Comment (markaj-nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:333]: Do we need this `else` branch?

File: src/platform/Zephyr/BLEManagerImpl.cpp (hunk 10 above)
Comment (LipinskiPNordicSemi) [reply src/platform/Zephyr/BLEManagerImpl.cpp:305]: Spec/safty isue.

File: src/include/platform/CHIPDeviceConfig.h (hunk 1 above)
Comment (LuDuda) [reply src/include/platform/CHIPDeviceConfig.h:696]: > The spec only mentions maximum duration of the announcement and that after 15 minutes it should switch to the extended one. Thus, a manufacturer can decide to advertise e.g. 1 hour if I read the spec correctly.

Okay, then it makes sense.

File: src/include/platform/CHIPDeviceConfig.h [hunk 12]
```diff
@@ -514,7 +514,7 @@
  *
  */
 #ifndef CHIP_DEVICE_CONFIG_ENABLE_PAIRING_AUTOSTART
-#define CHIP_DEVICE_CONFIG_ENABLE_PAIRING_AUTOSTART 1
+#define CHIP_DEVICE_CONFIG_ENABLE_PAIRING_AUTOSTART 0
```
Comment (Damian-Nordic) [inline src/include/platform/CHIPDeviceConfig.h:517]: Why this change?

File: src/platform/nrfconnect/CHIPDevicePlatformConfig.h [hunk 13]
```diff
@@ -270,6 +270,14 @@
 #define CHIP_DEVICE_CONFIG_ENABLE_EXTENDED_DISCOVERY 1
 #endif // CONFIG_CHIP_EXTENDED_DISCOVERY
 
+#ifdef CONFIG_CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
+#define CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING 1
+#endif // CONFIG_CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
+
+#ifdef CONFIG_CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_MINUTES
+#define CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_SECS CONFIG_CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_MINUTES * 60
+#endif // CONFIG_CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_MINUTES
```
Comment (Damian-Nordic) [inline src/platform/nrfconnect/CHIPDevicePlatformConfig.h:279]:
```suggestion
#define CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_SECS (CONFIG_CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_MINUTES * 60)
```
This is always defined for our and Zephyr platform.

File: src/platform/Zephyr/BLEManagerImpl.h [hunk 14]
```diff
@@ -91,6 +91,7 @@ class BLEManagerImpl final : public BLEManager, private BleLayer, private BlePla
         kAdvertisingRefreshNeeded =
             0x0010, /**< The advertising state/configuration has changed, but the SoftDevice has yet to be updated. */
         kChipoBleGattServiceRegister = 0x0020, /**< The system has currently CHIPoBLE GATT service registered. */
+        kExtendedAdvertisingEnabled  = 0x0040, /**< The appliaction enabled extended advertising*/
```
Comment (Damian-Nordic) [inline src/platform/Zephyr/BLEManagerImpl.h:94]:
```suggestion
        kExtendedAdvertisingEnabled  = 0x0040, /**< The application has enabled extended advertising. */
```

File: config/zephyr/Kconfig [hunk 15]
```diff
@@ -520,4 +520,21 @@ config CHIP_OTA_IMAGE_EXTRA_ARGS
...
 
+config CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
+	bool "Allow to extend advertising time longer than 15min"
+	help
+	  Enable BLE extended advertising.
+	  If this config is true, 
+	  CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_SECS can be set up to 48h.
+
+config CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_MINUTES
+	int "Configures commissioning tiemout"
```
Comment (Damian-Nordic) [inline config/zephyr/Kconfig:531]:
```suggestion
	int "BLE advertising duration"
```

File: config/zephyr/Kconfig [hunk 16]
```diff
@@ -520,4 +520,21 @@ config CHIP_OTA_IMAGE_EXTRA_ARGS
 
 endif
 
+config CHIP_DEVICE_CONFIG_BLE_EXT_ADVERTISING
+	bool "Allow to extend advertising time longer than 15min"
```
Comment (Damian-Nordic) [inline config/zephyr/Kconfig:524]:
```suggestion
	bool "BLE extended advertising (over 15 minutes)"
```

**Review by Damian-Nordic — APPROVED** [review]

File: src/include/platform/CHIPDeviceConfig.h (hunk 12 above)
Comment (markaj-nordic) [reply src/include/platform/CHIPDeviceConfig.h]: Reverted.

File: src/platform/nrfconnect/CHIPDevicePlatformConfig.h (hunk 13 above)
Comment (markaj-nordic) [reply src/platform/nrfconnect/CHIPDevicePlatformConfig.h]: Done.

**Review by Damian-Nordic — APPROVED** [review]

**Review by ArekBalysNordic — APPROVED** [review]
